libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/ChargingStationsMetadataUpdater.kt

```diff
@@ -0,0 +1,134 @@
+                    val updatedUnrecognizedProperties = waypointUnrecognizedProperties
+                        .toMutableMap()
+                    updatedUnrecognizedProperties["metadata"] = updatedMetadata
+                    waypoint.toBuilder()
+                        .unrecognizedJsonProperties(updatedUnrecognizedProperties)
+                        .build()
+                } else {
+                    waypoint
+                }
+            }
+        route.toBuilder().waypoints(updatedWaypoints).build()
+    }
+    val updatedResponse = response.toBuilder().routes(updatedRoutes).build()
+    return updatedResponse
+}
+
+private fun updateChargingStationsMetadataBasedOnRequestUrl(
+    rerouteRouteOptions: RouteOptions,
+    newPrimaryRoute: NavigationRoute
+): DirectionsResponse {
+    val chargingStationsPowers = rerouteRouteOptions
+        .getUnrecognisedSemicolonSeparatedParameter("waypoints.charging_station_power")
+    val chargingStationsIds = rerouteRouteOptions
+        .getUnrecognisedSemicolonSeparatedParameter("waypoints.charging_station_id")
+    val chargingStationsCurrentTypes = rerouteRouteOptions
+        .getUnrecognisedSemicolonSeparatedParameter("waypoints.charging_station_current_type")
+    val response = newPrimaryRoute.directionsResponse
+    val updatedRoutes = response.routes().map { route ->
+        val updatedWaypoints =
+            (route.waypoints() ?: emptyList()).mapIndexed { waypointIndex, waypoint ->
+                val chargingStationId = chargingStationsIds?.get(waypointIndex)
+                val chargingStationsPower = chargingStationsPowers?.get(waypointIndex)
+                    ?.toIntOrNull()
+                    ?.let { it / 1000 }
+                val chargingStationsCurrentType = chargingStationsCurrentTypes?.get(waypointIndex)
+                waypoint.toBuilder()
+                    .unrecognizedJsonProperties(
+                        mapOf(
+                            "metadata" to JsonObject().apply {
+                                if (chargingStationsPower != null) {
+                                    add("power_kw", JsonPrimitive(chargingStationsPower))
+                                }
+                                if (!chargingStationId.isNullOrEmpty()) {
+                                    add("station_id", JsonPrimitive(chargingStationId))
+                                }
+                                if (!chargingStationsCurrentType.isNullOrEmpty()) {
+                                    add(
+                                        "current_type",
+                                        JsonPrimitive(chargingStationsCurrentType)
+                                    )
+                                }
+                            }
+                        )
+                    )
+                    .build()
+            }
+        route.toBuilder().waypoints(updatedWaypoints).build()
```

Comment:
What about waypoints_per_route parameter? It may be false, then you should update the value in the response root, not per route.
I think it's allowed if alternatives are disabled.

Reply:
done